### ISSUE TITLE
Support for visitor.js and DIL,js functionality in scriptless scenarios

### ISSUE THREAD

**Issue by jmrodriguez-smartclip** [adobe/alloy#527]
[Note: i'm not sure if this is within the scope of the alloy project, because it should be solved served-side]

### Expected Behaviour
Although it looks strange to ask for a functionality currently given by js, when js is not available, this 
scenario is very usual for us. VAST (video) based advertising allows to set up URLs to be called when certain events happen (ad start, ad complete, ad clicked,etc), and we'd like to add users to segments based in those events.
Being just URLs, no script can be called at that time.What we'd expect, is some way to compose a single URL which will process the user side (visitor.js functionality), this is, checking an Adobe User Id cookie has been received, or setting it, and also, process the DMP side (DIL.js functionality), sending the signals especified in the URL, to the DMP.
Other parameters, like the GDPR consent string, etc, would be sent in that URL too.
### Actual Behaviour
Right now (and i mean, visitor.js/DIL.js), this is not possible.It's possible to send events (signals) to the DMP by mimicking the event request generated by DIL.js. But that only works for existing users, as they already have an Adobe User id cookie that it's sent along the event request.
There's no way to get a new user id and send signals to the DMP, in a single URL request.

**Comment by dancingcactus** [comment]
Just so I understand the idea would be to send all of the information as query parameters so that the URL could be called by an external system?  Do you typically have tokenization available in these systems (e.g. replace x query param with y value)?

**Comment by jmrodriguez-smartclip** [comment]
Yes, exactly that. Those systems support macro replacement, with a fixed set of possible macros, but those replacements mostly happen server side (before the VAST XML document is sent to the client). 
In the client, as those documents are requested and processed by the video player, we cant make assumptions on what a certain video player may support, or which macros it could replace. In any case, those replacements would be outside the VAST standard, so...we cannot rely on that.
Summing it up, the whole url for the DMP is built server side by the ad system, using macros. We cannot rely / dont have the chance of modifiying it client side.

**Comment by raresvla** [comment]
@jmrodriguez-smartclip We do have clear plans for two very common events:
* **clicks**, supporting click-trough tracking (collect and redirect)
* **impressions**, where we're simply going to collect the data

For these two, our current thinking is to have specialized endpoints (potentially `/click` and `/impression`), which will help with traffic management, observability and validation.

Are there domain-specific/standardized events that you'd like to be collected through this simpler interface, or you're rather looking for a generic `/collect` endpoint?

**Comment by dashiad** [comment]
Oops..I think i'm not giving enough context, so maybe the issue sounds strange.
We're not publishers, so we dont own the sites where our ads are shown. So we cannot add any javascript directly to the sites pages. So we cannot add Visitor.js or DIL.js at the page level.
In non-video ads, where the ads are served as html, we can add those scripts, and everything goes ok.
In video-ads, there are formats (VAST) where the ads are served as XML. Events are just URLs, so the ad has no way to insert a javascript in the page. It can call an endpoint (we're using /event, buy, yes, /click or /impressions would work) , but by using those endpoints, the signal will only work if the user that has received the ad, already had an Adobe Id. 

If the user doesnt already have an Adobe Id (and we cannot be sure about that, as we dont control the sites, and cant add the visitor.js javascript), the signal will be lost. 

As i see it, Adobe chose to create an API for shared services (Visitor.js), so others (DIL.js) could use them.And that makes sense. But we'll need some endpoint that does both jobs: this is: receive an event, check if the user id has been received, otherwise create it (and assign cookie), and then, do the same work that /event, /click or /impression would do.

So, it'd be like the very basic functionality of Alloy (creating users / signals) would be available 100% server side, and both of them could be triggered by a single url call.

**Comment by raresvla** [comment]
@dashiad Thank you for the details! That make sense and is exactly what `/click` and `/impression` (as well as the existing `/interact` and `/collect`) endpoints are supposed to do:
1. Make sure a device ID is present; if none is found on the request (either 1st party ID or a 3rd party cookie when enabled), the server will make sure to generate one
1. Transparently add the ID to the collected call so that a profile can be captured

With Alloy, there's no need for Visitor.js or DIL.js, all their functionality will be covered either by Alloy itself or by the server.

I do have a follow-up question on the identity topic: from the constraint environment of a video-player, what kind of identities will be available?
* Would that player be aware of logged in users, thus be able to provide that identity?
* Would it have access to device-level identities stored in 1st party cookies (in browsers)? What about 3rd party cookies (regardless of their very uncertain future)?
* How about mobile devices? Is there a use-case in which the video player is embedded in an application?

**Comment by jmrodriguez-smartclip** [comment]
Ooops, sorry for mixing up github identities (personal/job)
Ok, so, in the current version, the /collect endpoint already does that? I was unaware of that (i had asked about it to our tech contact, but the conclusion was that there was no such endpoint. The most similar documented use case i found was "pixel based data transfers: https://docs.adobe.com/content/help/en/audience-manager/user-guide/implementation-integration-guides/sending-audience-data/real-time-inbound-data-integration/pixel-based-data-transfer.html , and it uses the /event endpoint).

About players:
The prerrequisites for any of those three scenarios, would be:
- The player knows about things that should be explicit in the environment (logged user ids) or implicit in the environment (cookies, device ids,etc)
- The player provides a way to map those things, into the received VAST document (using some sort of macros).

There are many players out there, so those requisites should be honoured by any of them.
While it could be possible that req. 1 is supported by some players, req. 2 definitely isn't. It's possible that some players have that functionality, but it'd be outside the VAST standard, so not usable.

Those scenarios would be handled server-side:
- To request an ad, the publisher parametrizes the player with the url of the adserver. Requesting that URL will retrieve the ad (in XML format).
- The publisher could add parameters (key-values) to that url, so when the adserver receives them, replaces them in the VAST document, using macros (in the different impression, click, etc, event endpoints), before serving it.
- The VAST document is received by the players, and events, along with the replaced macros.

So, in the end, that information could also be stored in the DMP, but the player wouldnt have any role in the process.

**Comment by jmrodriguez-smartclip** [comment]
One small detail i forgot to mention: In this scenario, it should be possible to send privacy related params, as the TCF consent string, to be checked server side. 
Using the /event endpoint directly has the big problem that we dont know if the server would check it (i guess it wouldnt, as that check is done in the js libraries), so we've stopped using this endpoint, and the DMP,  completely for our video campaigns.

**Comment by raresvla** [comment]
Hi @jmrodriguez-smartclip!

Let me clarify a bit more the points that I've stated earlier:
> Ok, so, in the current version, the /collect endpoint already does that? I was unaware of that (i had asked about it to our tech contact, but the conclusion was that there was no such endpoint. The most similar documented use case i found was "pixel based data transfers: https://docs.adobe.com/content/help/en/audience-manager/user-guide/implementation-integration-guides/sending-audience-data/real-time-inbound-data-integration/pixel-based-data-transfer.html , and it uses the /event endpoint, so it wont assign new ids).

1. The existing `/collect` and `/interact` endpoints transparently generate an ID when none is present and then advance with the call processing. In the response, the edge service either:
   * sets a 1st party cookie ID, when the call was made from a browser in a 1st party context (on `adobe.smartclip.com`, for example)
   * sets a 3rd party cookie ID and instructs the SDK (Alloy or other) to write a 1st party ID as a 1st party cookie, when call was made from a browser in 3rd party context (mostly for advertising)
   * just instructs the SDK to persist the generated identity, when the call was not made from a browser (that accepts cookies).
1. As you can see, there's a contract between the caller and the edge service, which is in place to ensure the newly generated identity is persisted on the device.
1. **The current endpoints only accepts `POST` requests with JSON payload**, there's no `GET` endpoint yet.

--- 

> One small detail i forgot to mention: In this scenario, it should be possible to send privacy related params, as the TCF consent string, to be checked server side.

Yes. IAB standard will soon be supported by the `/collect` and `/interact` endpoints -- therefore, any new endpoint that we add to support sending data via simple URLs (`GET` requests) will inherit these capabilities. 

> Using the /event endpoint directly has the big problem that we dont know if the server would check it (i guess it wouldnt, as that check is done in the js libraries), so we've stopped using this endpoint, and the DMP, completely for our video campaigns.

The Adobe Audience Manager `/event` endpoint has support for IAB and does the proper check before collecting the data. It also forwards the received consent value when activating the data (i.e. sending it to data partners), per the IAB Standard requirements. You should reach back to your Adobe contact and ask for full details on this.

---

To sum up this thread, from our part:
1. We'll assess if `/click` and `/impression` are the only URL-based data collection endpoints that we want to expose, or if a generic one needs to be added as well
1. URL-based endpoints will have feature-parity with existing endpoints, meaning:
   1. transparent ID generation
   1. declared IDs support (explicit IDs from the environment, aka. logged-in users)
   1. IAB and other privacy standards support

One thing where I think we have to follow-up is 2.1 above, specifically to understand the given guarantees of having the newly generated ID persisted on the device.
